lib/tasks/uploads.rake

```diff
@@ -174,40 +230,156 @@ def migrate_from_s3(limit: nil)
         post.rebake!
-        putc "#"
+        posts_modified += 1
+        if max
+          # max is for debugging purposes, so be verbose
+          puts "\nModified #{posts_modified}/#{max}: #{post_tag}\n"
+        else
+          putc "#"
+        end
+        if max && posts_modified >= max
+          puts "Modification limit #{max} reached, halting migration"
+          return
+        end
+        # don't spam the default queue, limit rate of modification
+        wait_for_queue_tasks max_queue_tasks
       else
         putc "."
       end
 
-    rescue
-      putc "X"
+    rescue => e
+      # Unanticipated failure should be investigated before going further with a migration to limit harm
+      puts "Unhandled failure in #{post_tag} processing: #{e}"
+      exit 1
     end
   end
 
-  puts "Done!"
+  puts "All Post uploads migrated.  Migrating profile uploads..."
+
+  profiles_migrated = 0
+  migrate_profiles = UserProfile
+    .where("card_background_upload_id != 0 OR profile_background_upload_id != 0")
+  migrate_profiles = migrate_profiles.limit(limit) if limit
+
+  migrate_profiles.find_each do |profile|
+    puts "Migrating profile for #{profile.user.username} (#{profile.user.id})" if max
+    migrate_card = false
+    migrate_profile = false
+    if profile.card_background_upload&.url&.start_with?('//')
+      migrate_card = true
+    end
+    if profile.profile_background_upload&.url&.start_with?('//')
+      migrate_profile = true
+    end
+
+    if migrate_card
+      background_upload = profile.card_background_upload
+      puts "Card: #{background_upload.url}" if max
+      file = download_s3_upload(background_upload, max_file_size)
+      raise "#{url} not found for #{profile.user.name}" if file.nil?
+    end
+    if migrate_profile
+      profile_upload = profile.profile_background_upload
+      puts "Background: #{profile_upload.url}" if max
+      file = download_s3_upload(profile_upload, max_file_size)
+      raise "#{url} not found for #{profile.user.name}" if file.nil?
+    end
+
+    # These need to both be after the download but before setting new in case the same image was provided for both
+    profile.clear_card_background if migrate_card
+    profile.clear_profile_background if migrate_profile
+
+    if migrate_card
+      new_background = create_local_upload(background_upload, file, profile.user.id || -1)
+      profile.upload_card_background(new_background)
+    end
+    if migrate_profile
+      new_profile = create_local_upload(profile_upload, file, profile.user.id || -1)
+      profile.upload_profile_background(new_profile)
+    end
+
+    if migrate_card || migrate_profile
+      profile.save!
+      profiles_migrated += 1
+    end
+    if max && profiles_migrated >= max
+      puts "Modification limit #{max} reached, halting migration"
+      return
+    end
+    wait_for_queue_tasks max_queue_tasks
+  end
+
+  puts "All profiles uploads migrated.  Migrating other non-post uploads..."
```

Comment:
The full list (I hope it's full 😅) of non-post upload types is for example in the `ShrinkUploadedImage` class:

https://github.com/discourse/discourse/blob/630ec7b49cd98358736c5445302c509c62c68653/lib/shrink_uploaded_image.rb#L165-L173

This section of the script would break all those types of upload references (other than the profile uploads handled above)

Reply:
From meta:
> It feels like it makes sense to start with posts, since you want to rebake posts to change the URL in the cooked post after each migration; migrating all the uploads and then starting the rebakes would leave the site very broken for potentially a long time for a large site

You can process Uploads one by one, and rebake posts in-place, like [downsize_uploads](https://github.com/discourse/discourse/blob/630ec7b49cd98358736c5445302c509c62c68653/script/downsize_uploads.rb)/[ShrinkUploadedImage](https://github.com/discourse/discourse/blob/630ec7b49cd98358736c5445302c509c62c68653/lib/shrink_uploaded_image.rb) does.

---

Ideally we would extract the code from those two files into a class that'd be responsible for updating all objects related to the given upload (posts, user profiles, avatars, etc.)